File: stack/queueusingstack/MyQueue.java
package queueusingstack;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

class MyQueue {
    private Stack<Integer> inStack;  // Stack for push operations
    private Stack<Integer> outStack; // Stack for pop/peek operations

    // Constructor
    public MyQueue() {
        inStack = new Stack<>();
        outStack = new Stack<>();
    }

    // Push element x to the back of the queue
    public void push(int x) {
        inStack.push(x);
    }

    // Removes the element from in front of the queue and returns that element
    public int pop() {
        peek(); //or move()
        return outStack.pop();
    }

    // Get the front element
    public int peek() {
        move();
        return outStack.peek();
    }

    // Returns whether the queue is empty
    public boolean empty() {
        return inStack.isEmpty() && outStack.isEmpty();
    }

    // Transfer elements from inStack to outStack only when outStack is empty
    private void move() {
        if (outStack.isEmpty()) {
            while (!inStack.isEmpty()) {
                outStack.push(inStack.pop());
            }
        }
    }

    // Main method for testing the MyQueue class
    public static void main(String[] args) {
        Queue<Integer> q =  new LinkedList<Integer>();
        q.offer(1);
        q.offer(2);
        System.out.println("General Queue"+ q); //[1, 2]

        MyQueue myQueue = new MyQueue();
        myQueue.push(1); // Queue is [1]
        myQueue.push(2); // Queue is [1, 2]
        System.out.println(myQueue.peek()); // Output: 1
        System.out.println(myQueue.pop());  // Output: 1, Queue is now [2]
        System.out.println(myQueue.empty()); // Output: false
    }
}

File: stack/queueusingstack/MyQueue.py
class MyQueue:
    def __init__(self):
        # Stack to handle push operations
        self.inStack = []
        # Stack to handle pop and peek operations
        self.outStack = []

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.inStack.append(x)

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        self.move()
        return self.outStack.pop()

    def peek(self) -> int:
        """
        Get the front element.
        """
        self.move()
        return self.outStack[-1]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return not self.inStack and not self.outStack

    def move(self) -> None:
        """
        Transfer elements from inStack to outStack only when outStack is empty.
        """
        if not self.outStack:
            while self.inStack:
                self.outStack.append(self.inStack.pop())

# Example usage
myQueue = MyQueue()
myQueue.push(1)      # queue is [1]
myQueue.push(2)      # queue is [1, 2]
print(myQueue.peek()) # return 1
print(myQueue.pop())  # return 1, queue is [2]
print(myQueue.empty()) # return False
